src/vm/internal/otable.h

```diff
@@ -1,15 +1,51 @@
 #ifndef OTABLE_H
 #define OTABLE_H
 
-/*
- * otable.c
- */
-void init_otable(void);
-void enter_object_hash(struct object_t *);
-void remove_object_hash(struct object_t *);
-void remove_precompiled_hashes(char *);
-struct object_t *lookup_object_hash(const char *);
-int show_otable_status(struct outbuffer_t *, int);
-struct array_t *find_ch_n(const char *s);
+#include<cassert>
+#include<memory>
+#include<string>
+#include<unordered_map>
+#include<vector>
 
+#ifdef TESTING
+struct object_t
+{
+public:
+    std::string obname;
+
+    object_t(std::string const & s)
+    :obname(s) {}
+};
+#endif
+
+class ObjectTable /* final */
+{
+public:
+    using Key = std::string;
+    using Value = object_t*;
+    using Vector = std::vector<Value>;
+    using M1 = std::unordered_map<Key,Value>;
+    using M2 = std::unordered_map<Key,Vector>;
+    using Singleton = std::shared_ptr<ObjectTable>;
+
+    static Singleton get();
```

Comment:
Can you add some oneline document for each method?

Also, the entire impl seems to have no threading support, so it should be noted too.

Reply:
I think the driver is not rentrant in general so I am not sure about the MT comment. But sure I will add some comments.

Reply:
Added some comments to the functions to document them better.